[quant] Enable fusion for conv modules with bias

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36173 [quant] Enable fusion for conv modules with bias**

Summary:
Previously we were ignoring the conv bias during training if it existed
This PR adds the bias from the conv op during the conv+bn fusion process

Test Plan:
python test/quantization/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20921613](https://our.internmc.facebook.com/intern/diff/D20921613)

supriyar committed:
[quant] Enable fusion for conv modules with bias
Summary:
Previously we were ignoring the conv bias during training if it existed
This PR adds the bias from the conv op during the conv+bn fusion process

Test Plan:
python test/quantization/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -14,7 +14,7 @@ def __init__(self,
                  padding, dilation, transposed, output_padding,
                  groups,
-                 # bias: None, only support Conv with no bias
+                 bias,
```
We do not need bias as an argument here.

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -26,7 +26,7 @@ def __init__(self,
         nn.modules.conv._ConvNd.__init__(self, in_channels, out_channels, kernel_size,
                                          stride, padding, dilation, transposed,
-                                         output_padding, groups, False, padding_mode)
+                                         output_padding, groups, False if bias is None else True, padding_mode)
```
The underlying conv module does not contain bias, so this can be False always

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -96,17 +96,18 @@ def _forward(self, input):
-            conv = conv * rescale_factor.reshape([1, -1, 1, 1])
-            conv = conv + (self.beta - self.gamma * batch_mean * batch_rstd).reshape([1, -1, 1, 1])
-
+            conv = (self.gamma * batch_rstd).reshape([1, -1, 1, 1]) * conv_orig + \
```
The correct  equation is 
``` python
conv = (self.gamma * batch_rstd).reshape([1, -1, 1, 1]) * (conv_orig + self.bias -  batch_mean).reshape([1, -1, 1, 1])  + self.beta.reshape([1, -1, 1,1])
```

supriyar replied:
This would give incorrect results as batch_mean and batch_var need to take the bias into account

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -96,17 +96,18 @@ def _forward(self, input):
-            conv = conv + (self.beta - self.gamma * batch_mean * batch_rstd).reshape([1, -1, 1, 1])
-
+            conv = (self.gamma * batch_rstd).reshape([1, -1, 1, 1]) * conv_orig + \
+                (self.beta - self.gamma * batch_rstd * batch_mean).reshape([1, -1, 1, 1])
```
We need to also change the equation for the else condition (training is false to):
```
            conv = conv + self.gamma * (self.bias - self.running_mean)/running_std).reshape([1, -1, 1, 1])	 + self.beta.reshape([1, -1, 1,1])                
```

supriyar replied:
I don't think this needs to change since the bias is baked into `self.running_mean` and `running_std` after training

raghuramank100 replied:
That's not true, think of it this way: If I add a bias to my output during training, the running mean will be bias (assume that the conv output prior to bias add is zero mean). 
So, the output will be gamma/sigma * (y + bias- mean) + beta.
If you dont have the bias now the output will be:
gamma/sigma * (y - mean) + beta.
which will be incorrect.

supriyar replied:
Thanks for the explanation! Updated it.

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -26,7 +26,7 @@ def __init__(self,
+                                         output_padding, groups, False if bias is None else True, padding_mode)
         assert qconfig, 'qconfig must be provided for QAT module'
         self.qconfig = qconfig
         self.eps = eps
```
Create bias as a parameter based on constructor args:
```    
    if bias:
            self.bias = Parameter(torch.Tensor(out_channels))
        else:
            self.register_parameter('bias', None)
```

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -139,13 +140,13 @@ def from_float(cls, mod, qconfig=None):
         qat_convbn = cls(conv.in_channels, conv.out_channels, conv.kernel_size,
                          conv.stride, conv.padding, conv.dilation,
-                         conv.groups,
+                         conv.groups, conv.bias,
```
For clarity, might be better to explicitly make this a bool. i.e 
```
bias=conv.bias is not None
```

raghuramank100 commented on test/quantization/test_quantization.py:
```diff
@@ -1363,6 +1363,24 @@ def checkModelWithSequentialQuantized(self, model):
+        model = ModelForFusionWithBias().train()
+        model = fuse_modules(model, [["conv1", "bn1", "relu1"],
+                                     ["conv2", "bn2"]])
+        model.qconfig = default_qat_qconfig
```
This code only changes prepare_qat, so no need to test after convert. We should call prepare_qat with identity fake_quant, so that the numerics of the model are identical to the original model. 
```
model.qconfig = QConfig(activation=torch.nn.Identity,
                                      weight=torch.nn.Identity)
prep_model = prepare_qat(model, inplace=False)

supriyar committed:
Update on "[quant] Enable fusion for conv modules with bias"


Summary:
Previously we were ignoring the conv bias during training if it existed
This PR adds the bias from the conv op during the conv+bn fusion process

Test Plan:
python test/quantization/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Enable fusion for conv modules with bias"


Summary:
Previously we were ignoring the conv bias during training if it existed
This PR adds the bias from the conv op during the conv+bn fusion process

Test Plan:
python test/quantization/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantization.py:
```diff
@@ -1363,6 +1363,34 @@ def checkModelWithSequentialQuantized(self, model):
+        self.assertEqual(out_ref, out_fused)
+
+        model.qconfig = default_qat_qconfig
+        prepare_qat(model, inplace=True)
```
It looks like this part and computation of SQNR is not used, should it be removed?

raghuramank100 reviewed (approved):
Please see comments, thanks!

supriyar committed:
Update on "[quant] Enable fusion for conv modules with bias"


Summary:
Previously we were ignoring the conv bias during training if it existed
This PR adds the bias from the conv op during the conv+bn fusion process

Test Plan:
python test/quantization/test_quantization.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
